src/app/PathIterator.cpp

```diff
@@ -0,0 +1,206 @@
+extern uint8_t emberAfClusterIndex(EndpointId endpoint, ClusterId clusterId, EmberAfClusterMask mask);
+
+namespace chip {
+namespace app {
+
+void PathIterator::Reset(ClusterInfo * aClusterInfo)
+{
+    mpClusterInfo = aClusterInfo;
+
+    // Reset iterator state
+    mEndpointIndex  = UINT16_MAX;
+    mClusterIndex   = UINT8_MAX;
+    mAttributeIndex = UINT16_MAX;
+
+    Proceed();
+}
+
+void PathIterator::PrepareEndpointIndexRange(const ClusterInfo & aClusterInfo, uint16_t * aBeginEndpointIndex,
+                                             uint16_t * aEndEndpointIndex)
+{
+    if (!aClusterInfo.HasValidEndpointId())
+    {
+        *aBeginEndpointIndex = 0;
+        *aEndEndpointIndex   = emberAfEndpointCount();
+    }
+    else
+    {
+        *aBeginEndpointIndex = emberAfIndexFromEndpoint(aClusterInfo.mEndpointId);
+        // If the given cluster id does not exist on the given endpoint, it will return uint16(0xFFFF), then endEndpointIndex
+        // will be 0, means we should iterate a null endppint set (skip it).
+        *aEndEndpointIndex = static_cast<uint16_t>(*aBeginEndpointIndex + 1);
+    }
+}
+
+void PathIterator::PrepareClusterIndexRange(const ClusterInfo & aClusterInfo, EndpointId aEndpointId, uint8_t * aBeginClusterIndex,
+                                            uint8_t * aEndClusterIndex)
+{
+    *aBeginClusterIndex = 0;
+    *aEndClusterIndex   = emberAfClusterCount(aEndpointId, true /* server */);
+
+    if (aClusterInfo.HasValidClusterId())
+    {
+        for (uint8_t i = *aBeginClusterIndex; i < *aEndClusterIndex; i++)
+        {
+            ClusterId clusterId = emberAfGetNthClusterId(aEndpointId, i, true /* server */).Value();
+            if (clusterId == aClusterInfo.mClusterId)
+            {
+                *aBeginClusterIndex = i;
+                *aEndClusterIndex   = static_cast<uint8_t>(*aBeginClusterIndex + 1);
+                return;
+            }
+        }
+        *aBeginClusterIndex = UINT8_MAX;
+        *aEndClusterIndex   = 0;
+    }
+}
+
+void PathIterator::PrepareAttributeIndexRange(const ClusterInfo & aClusterInfo, EndpointId aEndpointId, ClusterId aClusterId,
+                                              uint16_t * aBeginAttributeIndex, uint16_t * aEndAttributeIndex)
+{
+    if (!aClusterInfo.HasValidAttributeId())
+    {
+        *aBeginAttributeIndex = 0;
+        *aEndAttributeIndex   = emberAfGetServerAttributeCount(aEndpointId, aClusterId);
+    }
+    else
+    {
+        *aBeginAttributeIndex = emberAfGetServerAttributeIndexByAttributeId(aEndpointId, aClusterId, aClusterInfo.mFieldId);
+        // If the given attribute id does not exist on the given endpoint, it will return uint16(0xFFFF), then endAttributeIndex
+        // will be 0
+        *aEndAttributeIndex = static_cast<uint16_t>(*aBeginAttributeIndex + 1);
+    }
+}
+
+bool PathIterator::Proceed()
+{
+    for (; mpClusterInfo != nullptr; (mpClusterInfo = mpClusterInfo->mpNext, mEndpointIndex = UINT16_MAX))
```

Comment:
I feel like this triple nested for loop, with two points in this function returning true, can be simplified to a single for loop.

Will need to mull on this.

Reply:
I feel the nested look makes the code a bit more easier to read. Since replace the return by continue willl make it just a normal iteration.